test/test_parametrized.py

```diff
@@ -0,0 +1,192 @@
+import torch
+from caustic import Simulator, EPL, Sersic, FlatLambdaCDM
+from caustic.utils import get_meshgrid
+
+
+def test_params():
+    # Test common simulator structure
+    class Sim(Simulator):
+        def __init__(self):
+            super().__init__()
+            self.cosmo = FlatLambdaCDM(h0=None)
+            self.epl = EPL(self.cosmo)
+            self.sersic = Sersic()
+            self.add_param("z_s", 1.0)
+    
+    sim = Sim()
+    assert len(sim.module_params) == 2 # dynamic and static
+    assert len(sim.module_params.dynamic) == 0 # simulator has no dynmaic params
+    assert len(sim.module_params.static) == 1 # and 1 static param (z_s)
+    assert len(sim.params) == 2 # dynamic and static
+    assert len(sim.params.dynamic) == 3 # cosmo, epl and sersic
+    assert len(sim.params.static) == 2 # simulator and cosmo have static params
+    assert len(sim.params.dynamic.flatten()) == 15 # total number of params
+
+
+def test_graph():
+    # Test common simulator structure
+    class Sim(Simulator):
+        def __init__(self, name="test_simulator"):
+            super().__init__(name)
+            self.cosmo = FlatLambdaCDM(h0=None, name="cosmo")
+            self.lens = EPL(self.cosmo, name="lens")
+            self.source = Sersic(name="source")
+            self.z_s = torch.tensor(1.0)
+            self.thx, self.thy = get_meshgrid(0.04, 20, 20)
+
+    sim = Sim()
+    sim.get_graph(True, True)
+    sim.get_graph()
+
+
+def test_unpack_all_modules_dynamic():
+    class Sim(Simulator):
+        def __init__(self, name="test"):
+            super().__init__(name)
+            self.cosmo = FlatLambdaCDM(h0=None, name="cosmo")
+            self.epl = EPL(self.cosmo, name="lens")
+            self.sersic = Sersic(name="source")
+            self.z_s = torch.tensor(1.0)
+            self.thx, self.thy = get_meshgrid(0.04, 20, 20)
+
+        def forward(self, params):
+            alphax, alphay = self.epl.reduced_deflection_angle(x=self.thx, y=self.thy, z_s=self.z_s, params=params) 
+            bx = self.thx - alphax
+            by = self.thy - alphay
+            return self.sersic.brightness(bx, by, params)
+
+    sim = Sim()
+
+    cosmo_params = [0.7]
+    lens_params = [0.5, 0., 0., 0.8, 0.5, 1.2, 1.1]
+    source_params = [0.0, 0.0, 0.8, 0.0, 1., 0.2, 10.]
+    
+    # test list input
+    x = [torch.tensor(_x) for _x in cosmo_params + lens_params + source_params]
+    sim.forward(sim.pack(x))
```

Comment:
Simulator object already calls pack on inputs so you can just do `sim(x)`

Also pack accepts list inputs, can't you just do: `sim(cosmo_params + lens_params + source_params)`?

Reply:
Yes I can, and what I have is essentially that. But x needed to be sent to tensor first so that I could test what happens. 

The way you constructed your example, I would have to send all lists (cosmo_params, lens_params and source_params) to tensor separately, which I simply did in a single line (line 65).